tests/kernel/update_kernel.pm

```diff
@@ -166,74 +162,70 @@ sub install_lock_kernel {
 }
 
 sub prepare_kgraft {
     my ($repo, $incident_id) = @_;
-    my $arch    = get_required_var('ARCH');
-    my $version = get_required_var('VERSION');
-    my $release_override;
-    my $lp_product;
-    my $lp_module;
-    if ($version eq '12') {
-        $release_override = '-d';
-    }
-    if (!is_sle('>=12-SP3')) {
-        $version = '12';
-    }
-    # SLE15 has different structure of modules and products than SLE12
-    if (is_sle('15+')) {
-        $lp_product = 'sle-module-live-patching';
-        $lp_module  = 'SLE-Module-Live-Patching';
-    }
-    else {
-        $lp_product = 'sle-live-patching';
-        $lp_module  = 'SLE-Live-Patching';
-    }
 
-    #install kgraft product
-    zypper_call("ar http://download.suse.de/ibs/SUSE/Products/$lp_module/$version/$arch/product/ kgraft-pool");
-    zypper_call("ar $release_override http://download.suse.de/ibs/SUSE/Updates/$lp_module/$version/$arch/update/ kgraft-update");
-    zypper_call("ref");
-    zypper_call("in -l -t product $lp_product", exitcode => [0, 102, 103]);
-    zypper_call("mr -e kgraft-update");
+    fully_patch_system;
 
     #add repository with tested patch
+    my $incident_klp_pkg;
     my @repos = split(",", $repo);
     while (my ($i, $val) = each(@repos)) {
-        zypper_call("ar $val kgraft-test-repo-$i");
-
-        my $kversion = zypper_search(q(-s -x kernel-default));
-        my $pversion = zypper_search("-s -t package -r kgraft-test-repo-$i");
-        $pversion = join(' ', map { $$_{name} } @$pversion);
-
+        my $cur_repo = "kgraft-test-repo-$i";
+        zypper_call("ar $val $cur_repo");
+        my $pkgs = zypper_search("-s -t package -r $cur_repo");
         #disable kgraf-test-repo for while
-        zypper_call("mr -d kgraft-test-repo-$i");
+        zypper_call("mr -d $cur_repo");
+
+        foreach my $pkg (@$pkgs) {
+            my $cur_klp_pkg = is_klp_pkg($pkg);
+            if ($cur_klp_pkg && $$cur_klp_pkg{kflavor} eq 'default') {
+                if ($incident_klp_pkg) {
+                    die "Multiple kernel live patch packages found: \"$$incident_klp_pkg{name}-$$incident_klp_pkg{version}\" and \"$$cur_klp_pkg{name}-$$cur_klp_pkg{version}\"";
+                }
+                else {
+                    $incident_klp_pkg = $cur_klp_pkg;
+                }
+            }
+        }
+    }
 
-        my $wanted_version = right_kversion($kversion, $pversion);
-        fully_patch_system;
-        install_lock_kernel($wanted_version);
+    if (!$incident_klp_pkg) {
+        die "No kernel livepatch package found";
+    }
 
-        if (check_var('REMOVE_KGRAFT', '1')) {
+    my $kversion       = zypper_search(q(-s -x kernel-default));
+    my $wanted_version = right_kversion($kversion, $incident_klp_pkg);
+    install_lock_kernel($wanted_version);
+
+    if (!check_var('REMOVE_KGRAFT', '1')) {
+        install_klp_product;
+    }
+    else {
+        while (my ($i, $val) = each(@repos)) {
```

Comment:
You don't need to cycle through the repos all over again. You can remove the packages in the first loop. Or you could just reuse the search results from the first loop here.

Reply:
Moving the REMOVE_KGRAFT code into the first loop would change semantics, because currently there's an install_lock_kernel invocation inbetween.

But I can add a patch implementing your second recommendation on reusing search results.

Reply:
Done. (Not in a separate patch though, because I realized I introduced that double loop. So I fixed it up inline)